text/0000-pi-types.md

```diff
@@ -0,0 +1,301 @@
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+We propose a simple, yet sufficiently expressive, addition of dependent-types
+(also known as, Π-types and value-types).
+
+Type checking will not require SMT-solvers or other forms of theorem provers.
+
+## Generic value parameters
+
+A `const` type parameter acts like a generic parameter, containing a constant
+expression. Declaring a generic parameter `const a: usize`, declares a constant
+variable `a` of type `usize`.
+
+One can create implementations, structs, enums, and traits, abstracting over
+this generic value parameter.
+
+Such a parameter acts type-like in the context of types, generics, and
+polymorphism, and value-like in the context of expressions, function bodies,
+and applications.
+
+## Compile time calculations on constant parameters
+
+Since it is simply consisting of constexprs, one can apply constant functions
+(`const fn`) to the parameter, to perform compile time, type level calculations
+on the parameter. This allows for great expressiveness as `const fn` improves.
+
+# Motivation
+[motivation]: #motivation
+
+An often requested feature is the "type-level numerals", which enables generic
+length arrays. The current proposals are often limited to integers or even lack
+of value maps, and other critical features.
+
+_Note:_ In an earlier version of this RFC, a `where` bound was proposed, but it
+proved too complex, so for now, types with invariants about the constant
+parameter can be used.
+
+It allows for creating powerful abstractions without type-level hackery.
+
+## What we want, and what we don't want
+
+We have to be very careful to avoid certain things, while still preserving the core features:
+
+1. Ability to use and manipulate values at type-level.
+2. The ability to use said values on expression-level (runtime).
+
+Yet, we do not want:
+
+1. SMT-solvers, due to not only undecidability (note, although, that SAT is
+   decidable) and performance, but the complications it adds to `rustc`.
+2. Monomorphisation-time errors, i.e. errors that happens during codegen of
+   generic functions. We try to avoid adding _more_ of these (as noted by
+   petrochenkov, these [already exists](https://github.com/rust-lang/rfcs/pull/1657#discussion_r68202733))
+
+# Detailed design
+[design]: #detailed-design
+
+## The new value-type construct, `const`
+
+Declaring a parameter `const x: T` allows using `x` in both an expression context
+(as a value of type `T`) and a type context (as a type parameter). In a sense,
+const "bridges" the world between values and types, since it allows us to
+declare value dependent types ([`ε → τ` constructors](https://en.wikipedia.org/wiki/Dependent_type)).
+
+Such a parameter is declared, like type parameters, in angle brackets (e.g.
+`struct MyStruct<const x: usize>`).
+
+The expr behavior is described as:
+
+    ValueParameterDeclaration:
+      Γ ⊢ const x: T
+      ──────────────
+      Γ ⊢ x: T
```

Comment:
I think what this is trying to say is that if the environment contains a `const x: T` declaration, `x` is of type `T`. That is, `x: T` in the premise has a different meaning from the `x: T` in the conclusion.

In general, I think the typing judgments in this RFC make it harder to follow.